### PR TITLE
Store ASG name in the resource CR

### PR DESCRIPTION
**Pull request by TheoBrigitte**
Towards: https://github.com/giantswarm/giantswarm/issues/5350

The `workerasgname` resource fetches the AutoScalingGroup name, on every resync period for every CR, this results in too many requests being issued to the AWS Cloudformation API, which then respond with `Throttling: Rate exceeded`.

In order to avoid this this PR update the `workerasgname` in order to fetch the ASG name once and then store it within the CR status, and re-use the CR status on next resync.

### CODE REVIEW

File: vendor/github.com/giantswarm/apiextensions/pkg/apis/provider/v1alpha1/aws_types.go [hunk 1]
```diff
@@ -178,6 +178,11 @@ type AWSConfigStatus struct {
 
 type AWSConfigStatusAWS struct {
 	AvailabilityZones []AWSConfigStatusAWSAvailabilityZone `json:"availabilityZones" yaml:"availabilityZones"`
+	AutoScalingGroup  AWSConfigStatusAWSAutoScalingGroup   `json:"autoScalingGroup" yaml:"autoScalingGroup"`
+}
+
+type AWSConfigStatusAWSAutoScalingGroup struct {
+	Name string `json:"name"`
```
Comment (TheoBrigitte) [inline vendor/github.com/giantswarm/apiextensions/pkg/apis/provider/v1alpha1/aws_types.go]: Note: vendor apiextensions using dep before merging.

File: service/controller/v23/resource/workerasgname/resource.go [hunk 2]
```diff
@@ -95,7 +132,23 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
...
+		customObject.Status.AWS.AutoScalingGroup.Name = workerASGName
+
+		_, err = r.g8sClient.ProviderV1alpha1().AWSConfigs(customObject.Namespace).UpdateStatus(&customObject)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", "updated CR status")
+
+		// TODO: test if this is needed.
```
Comment (xh3b4sd) [inline service/controller/v23/resource/workerasgname/resource.go]: It is when the object is changed.

Comment (TheoBrigitte) [reply service/controller/v23/resource/workerasgname/resource.go]: Good catch, I tested it and forgot to remove the TODO.

**Comment by TheoBrigitte** [comment]
vendoring of apiextensions done. Will merge on Monday.

TIL: we have dependencies to azure-sdk-go from aws-operator, weird

File: service/controller/v23/resource/workerasgname/resource.go [hunk 3]
```diff
@@ -42,16 +52,43 @@ func (r *Resource) Name() string {
...
+			return microerror.Mask(err)
+		}
+
+		newObj, err := r.g8sClient.ProviderV1alpha1().AWSConfigs(oldObj.GetNamespace()).Get(oldObj.GetName(), metav1.GetOptions{})
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		customObject = *newObj
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", "fetchted latest version of custom resource")
```
Comment (rossf7) [inline service/controller/v23/resource/workerasgname/resource.go]:
```suggestion
		r.logger.LogCtx(ctx, "level", "debug", "message", "fetched latest version of custom resource")
```

File: service/controller/v23/resource/workerasgname/resource.go [hunk 4]
```diff
@@ -42,16 +52,43 @@ func (r *Resource) Name() string {
...
 	}
 
 	controllerCtx, err := controllercontext.FromContext(ctx)
 	if err != nil {
 		return microerror.Mask(err)
 	}
 
+	{
+		r.logger.LogCtx(ctx, "level", "debug", "message", "finding the guest cluster worker ASG name in the CR")
```
Comment (rossf7) [inline service/controller/v23/resource/workerasgname/resource.go]:
```suggestion
		r.logger.LogCtx(ctx, "level", "debug", "message", "finding the tenant cluster worker ASG name in the CR")
```

Comment (rossf7) [reply service/controller/v23/resource/workerasgname/resource.go]: If its not lots of hassle I'd prefer to get the terminology updated.

**Comment by xh3b4sd** [comment]
I am not sure why this is not merged yet but I introduce a new version bundle. You need to move your changes then.

**Comment by TheoBrigitte** [comment]
It's not merged because half of e2e are failing so far, and it takes ages to go through it.

**Pull request merged**